### ISSUE TITLE
Relax version dependencies on CUDA pip wheels?

### ISSUE THREAD

**Issue by hawkinsp** [pytorch/pytorch#97990]
### 🐛 Describe the bug

A JAX user noted that `torch` and `jax` have conflicting requirements for NVIDIA's CUDA pip wheels.

Torch specifies exact version pins in its wheels:
```
Requires-Dist: nvidia-cuda-nvrtc-cu11 (==11.7.99) ; platform_system == "Linux" and platform_machine == "x86_64"
Requires-Dist: nvidia-cuda-runtime-cu11 (==11.7.99) ; platform_system == "Linux" and platform_machine == "x86_64"
Requires-Dist: nvidia-cuda-cupti-cu11 (==11.7.101) ; platform_system == "Linux" and platform_machine == "x86_64"
Requires-Dist: nvidia-cudnn-cu11 (==8.5.0.96) ; platform_system == "Linux" and platform_machine == "x86_64"
Requires-Dist: nvidia-cublas-cu11 (==11.10.3.66) ; platform_system == "Linux" and platform_machine == "x86_64"
Requires-Dist: nvidia-cufft-cu11 (==10.9.0.58) ; platform_system == "Linux" and platform_machine == "x86_64"
Requires-Dist: nvidia-curand-cu11 (==10.2.10.91) ; platform_system == "Linux" and platform_machine == "x86_64"
Requires-Dist: nvidia-cusolver-cu11 (==11.4.0.1) ; platform_system == "Linux" and platform_machine == "x86_64"
Requires-Dist: nvidia-cusparse-cu11 (==11.7.4.91) ; platform_system == "Linux" and platform_machine == "x86_64"
Requires-Dist: nvidia-nccl-cu11 (==2.14.3) ; platform_system == "Linux" and platform_machine == "x86_64"
Requires-Dist: nvidia-nvtx-cu11 (==11.7.91) ; platform_system == "Linux" and platform_machine == "x86_64"
```

These are non-overlapping with some of JAX's dependencies. JAX currently ships wheels that require:
* CUDA 11.8+ and cudnn 8.6+, or
* CUDA 12.0+ and cudnn 8.8+.

For CUDA 11 wheels, JAX requires:
```
          "nvidia-cublas-cu11>=11.11",
          "nvidia-cuda-nvcc-cu11>=11.8",
          "nvidia-cuda-runtime-cu11>=11.8",
          "nvidia-cudnn-cu11>=8.6",
          "nvidia-cufft-cu11>=10.9",
          "nvidia-cusolver-cu11>=11.4",
          "nvidia-cusparse-cu11>=11.7",
```

I'm wondering if you would be open to relaxing some or all of the CUDA version pins in `torch` to be `>=` constraints. NVIDIA promise backward compatibility for CUDA and CUDNN minor releases, so in theory this is safe. Theory and practice may differ, however.

My fall back is to tell users to do the following, which works because `pip` doesn't actually enforce the version constraints:
```
pip install torch
pip install --upgrade nvidia-cublas-cu11 nvidia-cuda-runtime-cu11 # ... etc.  
# then install JAX
```
but it's not optimal: it means one cannot really release a package that depends on both `jax` and `torch`.

### Versions

2.0

**Comment by ptrblck** [comment]
Based on your description it seems you are mixing the PyTorch wheels using CUDA 11.7 with Jax wheels using CUDA 11.8.
Our current build process for 11.8 ships dependencies inside the `torch` wheel, but @syed-ahmed is working on using the CUDA pip wheel dependencies for the 11.8 builds, too.
Once this is done, I would assume no conflicts would arise assuming you would install wheels with a matching CUDA version, which sounds a bit cleaner.
Would this work instead of relaxing the CUDA lib versions?

**Comment by hawkinsp** [comment]
No, I expect this will be an ongoing problem. Irrespective of *which* versions torch pins, a design where everyone pins *particular* versions will necessarily mean you cannot install multiple deep learning frameworks at the same time. Not unless we all decide to agree on supported versions somehow, and that seems unlikely.

**Comment by syed-ahmed** [comment]
The only way I can see multiple frameworks happy in the same machine is using virtualenv/pip installing inside conda/docker. I'm curious, what's your use case for installing multiple frameworks in the same python environment?

**Comment by malfet** [comment]
My concern with relaxed dependencies, is that it would be too easy to get in a situation when pytorch-X.Y is not technically not compatible with any of the latest cuda packages. (One can `pip install torch==1.13.1` and then try to update the libs under the hood to see what will happen) 
But if somebody at time of the release can guarantee, that `torch` will be compatible with range of package version, then I'm all for it

**Comment by hawkinsp** [comment]
@syed-ahmed Some example use cases: Colab notebooks, Kaggle notebooks, the deep learning VM images shipped by the various cloud vendors.

**Comment by ptrblck** [comment]
> No, I expect this will be an ongoing problem. Irrespective of which versions torch pins, a design where everyone pins particular versions will necessarily mean you cannot install multiple deep learning frameworks at the same time. 

The issue you are describing is cause since you are trying to install different frameworks with different minor CUDA versions instead of installing the matching one.
PyTorch supports CUDA 11.8 and builds the binaries for it, but you are still installing 11.7 into your Jax + CUDA 11.8 environment.

> Some example use cases: Colab notebooks, Kaggle notebooks, the deep learning VM images shipped by the various cloud vendors.

I'm still unsure why the matching versions are not used but based on your claim all these various cloud vendors want to mix different CUDA dependencies.

**Comment by hawkinsp** [comment]
@ptrblck I'm not personally trying to do anything: I'm one of the JAX maintainers trying to help users. In fact, I was trying to help some of your colleagues at NVIDIA who noted that it wasn't possible to `pip` install a package that depended on both `jax` and `torch`.

**Comment by Di-Is** [comment]
I want to use pytorch v2.1 and tensorflow v2.15.
I was agonizingly unable to resolve the nvidia-* dependency.

It is far from the best, I downloaded the wheel file, removed the nvidia-* dependency from the METADATA file, and installed the modified wheel.
